### PR TITLE
discrete ordinant, n-stream radiative transfer solver

### LINKED ISSUE #200: add n-stream radiative transfer solver
implement the n-stream, discrete ordinant radiative transfer solver

### PR DESCRIPTION
**Pull request by stacywalters**
closes #200

### CODE REVIEW

File: src/radiative_transfer/radiative_transfer.F90 [hunk 1]
```diff
@@ -266,30 +280,35 @@ subroutine mpi_unpack( this, buffer, position, comm )
...
+  
     call musica_mpi_unpack( buffer, position, this%n_streams_,  comm )
     call this%config_%mpi_unpack( buffer, position, comm )
 
-    call this%config_%get( "solver", solver, "", default="Delta Eddington" )
-    if( solver == 'discrete ordinants' ) then
-      call die_msg( 900569062,                                                &
-                    "Discrete ordinants method is not currently available" )
-    else
+    call config%get( "solver", solver, Iam, default="Delta Eddington" )
```
Comment (mattldawson) [inline src/radiative_transfer/radiative_transfer.F90:302]: Do we want this to default to Delta Eddington? Or should there be no default?

File: src/radiative_transfer/radiator.F90 [hunk 2]
```diff
@@ -15,9 +15,9 @@ module tuvx_radiator
   type radiator_state_t
     ! Optical properties for a radiator
 
-    real(kind=dk), allocatable :: layer_OD_(:,:) ! layer optical depth (vertical layer, wavelength bin)
+    real(kind=dk), allocatable :: layer_OD_(:,:)  ! layer optical depth (vertical layer, wavelength bin)
     real(kind=dk), allocatable :: layer_SSA_(:,:) ! layer single scattering albedo (vertical layer, wavelength bin)
-    real(kind=dk), allocatable :: layer_G_(:,:) ! layer asymmetry factor (vertical layer, wavelength bin)
+    real(kind=dk), allocatable :: layer_G_(:,:,:) ! layer asymmetry factor (vertical layer, wavelength bin)
```
Comment (mattldawson) [inline src/radiative_transfer/radiator.F90:20]: what should we label the third dimension as?

File: src/radiative_transfer/radiator.F90 [hunk 3]
```diff
@@ -324,35 +324,54 @@ subroutine accumulate( this, radiators )
...
+                                      + G(:,:,1)**i_stream * dscat
+          end do
+        else
+        ! rayleigh scattering
+          if( n_streams >= 2 ) then
+            asym_accum(:,:,2)  = asym_accum(:,:,2) + kair_asym_factor * dscat
+          else
+            asym_accum(:,:,1)  = asym_accum(:,:,1) + G(:,:,1) * dscat
+          end if
+        end if
```
Comment (mattldawson) [inline src/radiative_transfer/radiator.F90:374]: we should add to the function description how air is treated differently

File: src/radiative_transfer/radiator.F90 (hunk 2 above)
Comment (mattldawson) [reply src/radiative_transfer/radiator.F90:20]: we should also add to the function description how this is treated in this function

File: src/radiative_transfer/solvers/CMakeLists.txt [hunk 4]
```diff
@@ -2,6 +2,8 @@
 # Radiative transfer solver source
 
 set(SRC delta_eddington.F90
+        disord_subs.F90
```
Comment (mattldawson) [inline src/radiative_transfer/solvers/CMakeLists.txt:5]: maybe: `discrete_ordinant_util.F90`?

**Comment by stacywalters** [comment]
The third dimension is the number of radiation streams.  For
delta-eddington n_streams = 1, whereas for
the discrete ordinant solver n_streams = 2*i, i = 1,16.

**Comment by stacywalters** [comment]
I can't really help much, you would need input from a scientist, best from
Sasha Madronich.

**Comment by stacywalters** [comment]
Not sure what to say, the discrete_ordinant.F90 and disord_subs.F90 source
files comprise the
discrete ordinant radiative transfer solver.

**Review by mattldawson — APPROVED** [review]

**Pull request merged**